src/common/code_environment.cc

```diff
@@ -79,6 +79,34 @@ int get_process_name(char *buf, int len)
   return 0;
 }
 
+#elif defined(_WIN32)
+
+int get_process_name(char *buf, int len)
+{
+  if (len <= 0) {
+    return -EINVAL;
+  }
+
+  int length = GetModuleFileNameA(nullptr, buf, len);
+  if (length <= 0)
+    return -ENOSYS;
+  if (length == len)
+    buf[length] = '\0';
+
+  char* start = strrchr(buf, '\\');
+  if (!start)
+    return -ENOSYS;
+  start++;
+  char* end = strstr(start, ".exe");
+  if (!end)
+    return -ENOSYS;
+  *end = '\0';
+
+  strncpy(buf, start, len - 1);
```

Comment:
sorry, i didn't realize that you want to trim off the beginning bits and the extension before leaving my first comment. but since the buffer should be large enough under most circumstances, i think it's fine to reuse it.

i don't think `strncpy()` allows overlapped src and dest buffer, could you use `memmove()` instead ? like
```c++
memmove(buf, start, end - start);
buf[end - start] = '\0';
```
and we can the drop the previous `*end = '\0';`


what do you think?

Reply:
Originally I was using a temporary buffer and forgot that `strncpy()` is unsafe for overlapped source and destination.
You are correct,  we should use `memmove()`.
Thanks a lot for the pointers and suggestion!